src/main/java/com/google/devtools/build/lib/sandbox/SandboxHelpers.java

```diff
@@ -518,6 +543,18 @@ public SandboxInputs processInputFiles(
                 RootedPath.toRootedPath(
                     sourceRootToSandboxSourceRoot.get(sourceRoot),
                     inputArtifact.getRootRelativePath());
+          } else if (!inputArtifact.isSourceArtifact() && sandboxSourceRoots != null) {
+            FileArtifactValue metadata = inputMetadataProvider.getInputMetadata(actionInput);
+            if (!metadata.isRemote() && metadata.getMaterializationExecPath().isPresent()) {
+              inputPath = processResolvedSymlink(
```

Comment:
@lberki I think that the missing case that is causing https://github.com/bazelbuild/bazel/issues/20621 is that of a symlink to a source artifact. Since `processResolvedSymlink` doesn't take the source root mapping into account, it will return those symlinks unchanged, which is not correct.

I think that we would need to exercise the same logic as in L534-L545, but only if the symlink resolves to a source artifact. I am not sure how to check for the latter.

Reply:
Ah, I guess we can just check whether the resolved path lies under the exec root. If you don't get to work on this before the holidays, I can give it a try tomorrow.

Reply:
You're absolutely right and in fact I discovered this independently due to some test breakages (our tests are winning!)-- by the time you wrote this, I had the fix. I pushed the latest version to https://github.com/bazelbuild/bazel/pull/20603 . It works for me (protobuf, source tree + output base under `/tmp`, no disk cache). Mind also trying?

(don't look at the source code if you value your eyes, it's "hack to get all test cases green" quality and I'm not sure if all test cases are actually green)

Reply:
+1 for the test coverage, this is actually working now. I won't give a +1 for the control flow, but it's getting the job done ;⁠-⁠)

Reply:
Yeah, I'm kinda embarrassed having it on the public Internet, even in a pull request :)